kopf/reactor/causation.py

```diff
@@ -0,0 +1,132 @@
+"""
+Detection of the event causes, based on the resource state.
+
+The low-level watch-events are highly limited in information on what
+caused them, and they only notify that the object was changed somehow:
+
+* ``ADDED`` for the newly created objects (or for the first-time listing).
+* ``MODIFIED`` for the changes of any field, be that metainfo, spec, or status.
+* ``DELETED`` for the actual deletion of the object post-factum.
+
+The conversion of the low-level *events* to the high-level *causes* is done by
+checking the object's state and comparing it to the saved last-seen state.
+
+This allows to track which specific fields were changed, and are those changes
+important enough to call the handlers: e.g. the ``status`` changes are ignored,
+so as some selected system fields of the ``metainfo``.
+
+For deletion, the cause is detected when the object is just marked for deletion,
+not when it is actually deleted (as the events notify): so that the handlers
+could execute on the yet-existing object (and its children, if created).
+"""
+import logging
+from typing import NamedTuple, Text, Mapping, MutableMapping, Optional, Any, Union
+
+from kopf.reactor import registries
+from kopf.structs import diffs
+from kopf.structs import finalizers
+from kopf.structs import lastseen
+
+# The constants for the event types, to prevent the direct string usage and typos.
+# They are not exposed by the framework, but are used internally. See also: `kopf.on`.
+NEW = 'new'
+CREATE = 'create'
+UPDATE = 'update'
+DELETE = 'delete'
+NOOP = 'noop'
+FREE = 'free'
+GONE = 'gone'
+
+# The human-readable names of these causes. Will be capitalised when needed.
+TITLES = {
+    CREATE: 'creation',
+    UPDATE: 'update',
+    DELETE: 'deletion',
+}
+
+
+class Cause(NamedTuple):
+    """
+    The cause is what has caused the whole reaction as a chain of handlers.
+
+    Unlike the low-level Kubernetes watch-events, the cause is aware
+    of the actual field changes, including the multi-handlers changes.
+    """
+    logger: Union[logging.Logger, logging.LoggerAdapter]
+    resource: registries.Resource
+    event: Text
+    body: MutableMapping
+    patch: MutableMapping
+    diff: Optional[diffs.Diff] = None
+    old: Optional[Any] = None
+    new: Optional[Any] = None
+
+
+def detect_cause(
+        event: Mapping,
+        **kwargs
+) -> Cause:
+    """
+    Detect the cause of the event to be handled.
+
+    This is a purely computational function with no side-effects.
+    The causes are then consumed by `custom_object_handler`,
+    which performs the actual handler invocation, logging, patching,
+    and other side-effects.
+    """
+    etyp = event['type']  # e.g. ADDED, MODIFIED, DELETED.
+    body = event['object']
+
+    # The object was really deleted from the cluster. But we do not care anymore.
+    if etyp == 'DELETED':
```

Comment:
maybe "event_type" instead of "etyp"? I don't like abbreviations/acronyms :-/

Reply:
Fixed by removing the variable completely. It is used only in one place now (previously (long time ago) there were 2-3 places).